sched/adaptdl_sched/policy/pollux.py

```diff
@@ -311,17 +345,28 @@ def _mutation(self, states, **kwargs):
         prob = 1.0 / np.where(states > 0, num_nonzero, num_zero)
         prob = prob.reshape(states.shape)
         m = np.random.random(states.shape) < prob
-        r = np.random.randint(np.iinfo(np.int16).max, size=states.shape)
+        r = np.random.randint(self._min_replicas, self._max_replicas + 1,
+                              size=states.shape)
         states = (states + m * r) % (self._max_replicas + 1)
+        # Assign minimum allowable replicas for jobs
+        min_replicas_idx = [i for i, j in enumerate(self._jobs)
+                            if j.min_replicas > 0]
+        states[:, min_replicas_idx] = self._min_replicas[min_replicas_idx]
         return states.reshape(states.shape[0], -1)
 
     def _repair(self, pop, **kwargs):
         states = pop.get("X")
         states = states.reshape(states.shape[0], *self._base_state.shape)
-        # Enforce at most one distributed job per node.
+        # Copy previous allocations for non-preemptible jobs
+        states[:, self._nonpreemptible_indices] = \
+            self._base_state[self._nonpreemptible_indices, :]
+        # Enforce at most one distributed job per node. Exclude all
+        # nonpreemptible jobs.
         distributed = np.count_nonzero(states, axis=2) > 1
         mask = states * np.expand_dims(distributed, axis=-1) > 0
         mask = mask.cumsum(axis=1) > 1
+        np_indices = [i for i, j in enumerate(self._jobs) if not j.preemptible]
+        mask[:, np_indices] = False
```

Comment:
I think this code subtly relies on the ordering of non-preemptible jobs being before preemptible jobs in the `states`, which is brittle and can break if this behavior is changed in the future.

For example, consider if a distributed pre-emptible job A comes before non-preemptible job B, and A and B share a node. Initially, the mask for A will be 0 and B will be 1. Then, this line will reset the mask for B to 0, resulting in preemptible job A sharing the node with non-preemptible job B.

Reply:
You are right. I don't think there is good way to enforce this because we are treating both NP and P jobs equally until they get an allocation.

Reply:
removed